hw/xbox/xid-sdl.c

```diff
@@ -464,6 +446,81 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+static int apply_haptic_effect(USBXIDState *s, SDL_HapticEffect *effect) {
+    int effect_id = SDL_HapticNewEffect(s->sdl_haptic, effect);
+    if (effect_id == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to create haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    int run_result = SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+    if (run_result == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to run haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    s->sdl_haptic_effect = effect;
+    s->sdl_haptic_effect_id = effect_id;
+    return 0;
+}
+
+static int haptic_effect_init(USBXIDState *s, SDL_HapticEffect *effect) {
+    if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) {
+        effect->leftright = (SDL_HapticLeftRight){
+            .type = SDL_HAPTIC_LEFTRIGHT,
+            .length = SDL_HAPTIC_INFINITY,
+            .large_magnitude = 0x0000,
+            .small_magnitude = 0x0000,
+        };
+
+        return apply_haptic_effect(s, effect);
+
+    } else if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_SINE) {
+        fprintf(stderr, "Joystick doesn't support left/right haptic feedback, emulating with SINE");
+
+        effect->periodic = (SDL_HapticPeriodic){
+            .type = SDL_HAPTIC_SINE,
+            .period = 1000,
+            .magnitude = 0,
+            .length = SDL_HAPTIC_INFINITY,
+            .attack_length = 0,
+            .fade_length = 0,
+            .delay = 0,
+        };
+
+       
+        return apply_haptic_effect(s, effect);
+    } else {
+        fprintf(stderr, "Joystick doesn't support necessary haptic effect\n");
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        return -1;
+    }
+}
+
+static int haptic_init(USBXIDState *s) {
+    if (SDL_HapticOpened(s-> device_index)) { return 0; }
+
+    s->sdl_haptic = SDL_HapticOpen(s->device_index);
+    if (s->sdl_haptic == NULL) {
+        fprintf(stderr, "Joystick doesn't support haptic effects\n");
+        return -1;
+    }
+
+    SDL_HapticEffect *effect = g_malloc(sizeof(SDL_HapticEffect));
+    int result = haptic_effect_init(s, effect);
+    if (result == -1) { g_free(effect); }
```

Comment:
`haptic_effect_init` re-initializes the effect struct  from scratch anyway.

There's no reason to `malloc` it. Just use a local variable here (unless SDL requires it to remain in memory - which is unlikely).